xarray/core/variable.py

```diff
@@ -3033,3 +3020,43 @@ def assert_unique_multiindex_level_names(variables):
                     "conflicting level / dimension names. {} "
                     "already exists as a level name.".format(d)
                 )
+
+
+def propagate_attrs_encoding(
+    old_variables: Mapping[Any, Variable], new_variables: Mapping[Any, Variable]
+) -> None:
+    """Propagate any attrs and/or encoding items from old variables that are not present
+    in new variables.
+
+    """
+    for name, var in new_variables.items():
+        old_var = old_variables.get(name)
+        if old_var is not None:
+            var.attrs = {**old_var.attrs, **var.attrs}
+            var.encoding = {**old_var.encoding, **var.encoding}
```

Comment:
I wonder if we should do a shallow copy of each new variable instead of modifying them directly, and return that? I usually try to avoid functions with side-effects as much as possible...

Reply:
I removed that function, which was only used within `Dataset._overwrite_indexes`. I still update the input `variables` in the latter instead of making a shallow copy because those variables won't be reused elsewhere and also because copying indexed coordinate variables do not currently preserve the identity of the wrapped pandas indexes (in this refactor we rely on this identify to find the indexes and their corresponding coordinates). This would need a deeper refactor of `IndexVariable.copy()`...